research/conditional/moe_layers/moe_gating.py

```diff
@@ -59,32 +61,54 @@ def __init__(
         assert not self.softmax_ungrouped or self.group_by_batch
 
     def calculate_gate(self, x, batch_size, seq_len):
-        with measure_time(self, "expert_embedding"):
-            if self.use_torch_bmm:
-                gate = self.get_gate().unsqueeze(0).expand(batch_size, -1, -1)
-                gate_out = torch.bmm(x, gate).permute(2, 0, 1)
-                assert gate_out.shape == (self.n_experts, batch_size, seq_len)
-            else:
+        if self.get_router_values_from == "ground_truth_weightless":
+            with torch.no_grad():
+                assert x.shape == (batch_size, seq_len, self.dmodel)
+                # memory error, to large tensor i think :(
+                # tokens_for_all_experts = x.reshape(
+                #     1, batch_size * seq_len, self.dmodel
+                # ).repeat(self.n_experts, 1, 1)
+                # experts_output = self.expert_inner_function(tokens_for_all_experts).to(
+                #     x.dtype
+                # )
+                tokens = x.reshape(1, batch_size * seq_len, self.dmodel).repeat(
+                    self.n_experts, 1, 1
+                )
                 gate_out = einsum(
-                    "batch_size seq_len dmodel, dmodel n_experts "
-                    "-> n_experts batch_size seq_len ",
-                    x,
-                    self.get_gate(),
+                    "n_experts b_s dmodel, n_experts dmodel dff -> n_experts b_s",
+                    tokens,
+                    self.expert_inner_function.lin1_weight,
                 )
-        # each expert chooses k within dimension 1
-        if not self.group_by_batch and not self.softmax_ungrouped:
-            gate_out = gate_out.reshape(self.n_experts, batch_size * seq_len)
-        # perform softmax either over tokens for each expert or over experts for each token
-        with measure_time(self, "softmax"):
-            if self.softmax_over == "tokens":
-                gate_out = torch.softmax(gate_out, dim=1)
-            elif self.softmax_over == "experts":
+
                 gate_out = torch.softmax(gate_out, dim=0)
-        if self.softmax_ungrouped:
-            gate_out = gate_out.reshape(self.n_experts, batch_size * seq_len)
 
-        if self.moe_values_exp != 1.0 or not isinstance(self.moe_values_exp, float):
-            gate_out = gate_out**self.moe_values_exp
+        else:
+            with measure_time(self, "expert_embedding"):
+                if self.use_torch_bmm:
```

Comment:
tutaj w tym ifie dodaj jedna linia
```
 if  self.get_router_values_from == "ground_truth_weightless"
  gate_out = einsum(
                      "batch_size seq_len dmodel, n_experts dmodel dff "
                      "-> n_experts batch_size seq_len dff ",
                      x,
                      self.expert_inner_function.lin1_weight,
                  ).sum(-1)
elif self.use_torch_bmm:
```
(chyba ze to to samo?)

Reply:
to w zasadzie to samo, tylko wymiary na wyjściu inne. dodałem jeszcze aktywację